### PR TITLE
Implementing Ditto for CHIL paper experimentation

### PR DESCRIPTION
**Pull request by emersodb**
# PR Type
Feature

# Short Description

Implementing Ditto as an additional baseline method for FL4Health, but also to add to the CHIL paper results.

# Tests Added

Basic tests verifying functionality for Ditto have been implemented.

### CODE REVIEW

File: examples/ditto_example/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# FedProx Federated Learning Example
+
+In this example, we implement the Ditto algorithm from the paper [Ditto: Fair and Robust Federated Learning Through Personalization
+](https://arxiv.org/abs/2012.04221). This method is has been shown to perform quick well on certain benchmarks (though not as well as others). See for example [Benchmark for Personalized Federated Learning](https://www.computer.org/csdl/journal/oj/2024/01/10316561/1S2UbvQk5Tq) and [pFL-Bench: A Comprehensive Benchmark for Personalized Federated Learning](https://arxiv.org/abs/2206.03655). The method is somewhat related to FedProx. Essentially, the model trains a global model through FedAvg and uses that global model, at each round to also train a personalized model. That model is trained using the same loss function, but with the initial global model (at the start of each round) as a penalty. That is, the local model weights are constrained using $\lambda > 0$ such that
```
Comment (fatemetkl) [inline examples/ditto_example/README.md:4]: is has -> has

**Review by fatemetkl — APPROVED** [review]
Looks good to me, just had some minor comments! Thanks for adding this so quickly!

File: fl4health/clients/ditto_client.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+            metrics (Sequence[Metric]): Metrics to be computed based on the labels and predictions of the client model
+            device (torch.device): Device indicator for where to send the model, batches, labels etc. Often 'cpu' or
+                'cuda'
+            loss_meter_type (LossMeterType, optional): Type of meter used to track and compute the losses over
+                each batch. Defaults to LossMeterType.AVERAGE.
+            checkpointer (Optional[TorchCheckpointer], optional): Checkpointer to be used for client-side
+                checkpointing. Defaults to None.
+            metrics_reporter (Optional[MetricsReporter], optional): A metrics reporter instance to record the metrics
+                during the execution. Defaults to an instance of MetricsReporter with default init parameters.
+            lam (float, optional): weight applied to the Ditto drift loss. Defaults to 0.0.
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/clients/ditto_client.py:45]: Fixed 🙂

File: fl4health/clients/ditto_client.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+
+    def compute_training_loss(
+        self,
+        preds: Dict[str, torch.Tensor],
+        features: Dict[str, torch.Tensor],
+        target: torch.Tensor,
+    ) -> TrainingLosses:
+        """
+        Computes training losses given predictions of the global and local models and ground truth data.
+        For the local model we adds to vanilla loss function by including Ditto penalty loss which is the l2 inner
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/clients/ditto_client.py:281]: Fixed 🙂

File: examples/ditto_example/README.md (hunk 1 above)
Comment (emersodb) [reply examples/ditto_example/README.md:4]: Fixed 🙂

**Comment by jewelltaylor** [comment]
@emersodb @fatemetkl Just got through the Ditto paper, will have feedback by 6:30 EST. Just going through now.

File: fl4health/clients/ditto_client.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+        assert len(self.initial_global_tensors) == len(local_model_weights)
+        assert len(self.initial_global_tensors) > 0
+
+        layer_inner_products: List[torch.Tensor] = [
+            torch.pow(torch.linalg.norm(initial_layer_weights - iteration_layer_weights), 2.0)
+            for initial_layer_weights, iteration_layer_weights in zip(self.initial_global_tensors, local_model_weights)
+        ]
+
+        # network l2 inner product tensor weighted by lambda
+        return self.lam * torch.stack(layer_inner_products).sum()
```
Comment (jewelltaylor) [inline fl4health/clients/ditto_client.py:111]: Should this be self.lam/2 to be consistent with paper (and [implementation](https://github.com/litian96/ditto/blob/26d9b29a62f3872dcdf071be08e897607c283bb4/flearn/trainers_MTL/ditto.py#L117))? Without it we would be scaling the contribution of the regularization by 2 which isn't a huge deal but if we are using the paper to get the space of lam to iterate over as hyperparam we should scale the loss or lams accordingly.

Comment (emersodb) [reply fl4health/clients/ditto_client.py:111]: 100%, that was an oversight by me. Will update this and the tests accordingly.

Comment (jewelltaylor) [reply fl4health/clients/ditto_client.py:111]: I can do it if you like! Don't want to make you do it on vacation haha! Just let me know

Comment (emersodb) [reply fl4health/clients/ditto_client.py:111]: Already in progress 🙂

Comment (jewelltaylor) [reply fl4health/clients/ditto_client.py:111]: Other than that, everything looks good to me! So I will take a look at the changes quickly and we can merge it in to get the ball rolling for experiments

Comment (emersodb) [reply fl4health/clients/ditto_client.py]: Sounds good. Make sure you take a little bit of time to verify the FLamby scripts in the other PR. I've gone over them a few times, but it helps to have a fresh set of eyes, just in case I've gone blind to anything.

Comment (jewelltaylor) [reply fl4health/clients/ditto_client.py]: Yeah I will definitely spend some time on the next PR as well! Any feedback should come in the next few hours :)

**Review by jewelltaylor — APPROVED** [review]
LGTM!
